### DISCUSSION TITLE
Vault events

### DISCUSSION THREAD

**Discussion by stephenwf** [digirati-co-uk/iiif-canvas-panel#92] [Ideas]
At the moment Vault is a way to store and normalise IIIF. It provides ways to access and also update that IIIF. It also handles some state for tracking which remote resources have been fetched. 

A recent expansion to this model is to associate "meta" information about IIIF resources, that are not a part of the IIIF itself, but are tied to them. Some examples of this "meta" information:

* Auth lifecycle state
* Chosen choices (for annotations)
* Enabled layers, opacity of those layers
* CORS support on image resource, for making decisions on rendering
* Event handlers on IIIF resources, possibly before they are rendered

This further tried to disconnect data from display and functionality. For example, this would allow us to build all of the workflows and mechanisms for an Auth-lifecycle without prescribing a user interface or library. This should also not be an inaccessible feature of the Vault, instead a feature that can be used for custom applications but also sharable.

An example of a **custom implementation** may be tracking diffs/changes in an editing scenario where a simple before/after is not enough information to save those changes. It could be a "last updated" timestamp associated with a IIIF resource.

Canvas panel or Atlas don't need to know or react to all of this information. But it can use this meta mechanism for some features.

### Vault events

The following example from the documentation site lead to this development. I'll break down the thought process here:
```js
cp.query({
  type: "Annotation", // `DisplayAnnotation` here?
  motivation: "linking",
}).addEventListener("click", (e) => {
  // Handle the event here.
});
```

Breaking these down into 2 steps - querying and event binding. To work in the Vault model, we need to ensure that these APIs:

* **Resistant to change** - if you load a new "linking" annotation, it should also get the event listener
* **Optional UI** - the events may be bound before the UI is rendered, or the UI may change. Event listeners should be be persistent regardless of where they are used.
* **Reusable** - If you create supporting meta for IIIF resources, you should have a path to reuse it and share it. 

So the code snippet above, we have 2 mechanisms. We have a query, that could change over time with items added or removed, and then we have an event bound - where the UI may not yet exist. So what this code snippet is saying: **Until I say otherwise, any annotation with a motivation linking annotation should have this click event**

So the logical place to store this is close to the data. For an individual annotation, an even listener would be stored like this in the Vault:
```js
  const state = {
    meta: {
      'https://example.org/annotation-1': {
        eventManager: {
          onClick: [
            { handler: () => {/* ... */ } },
          ],
        },
      },
    },
  };
```

Vault handles registering and unregistering this handler, and UI components can get a Reactive hook on these events when they are ready to render. 

In terms of usability, in order to create the above state from Vault you would write:
```js
const annotation = { type: 'Annotation', id: 'https://example.org/annotation-1' };

const listener = vault.addEventListener(annotation, 'onClick', () => { 
 // ...
});

// ...
vault.removeEventListener(annotation, 'onClick', listener);
```

These building blocks then are then bundled into easy to use tools in frameworks, such as React.

```jsx
function SomeComponent({ annotation }) {
  // Register an event listener with a hook somewhere in the application.
  useEventListener(annotation, 'onClick', (e: any) => {
    console.log(annotation);
  });

  return <App />
}

function RenderMyAnnotation({ annotation }) {
  const eventProps = useResourceEvents(annotation);

  // When you render your annotation, bind any events.
  return <div {...eventProps} />
}
```

Since these tools we create are largely inaccessible to pass new events or props down, this approach allows for the Vault to be set up before getting to a self-contained application. So for example:

```js
const vault = new Vault();

const manifest = await vault.loadManifest( ... );

vault.addEventListener(manifest.items[0], 'onClick', () => {
  // Do something when clicking on the first canvas.
});

renderMyComponent();
```

And then the component would have access to that event listener already.

**Comment by stephenwf** [comment]
Will follow this up with how the `meta` will be used to speed up querying with indexes. That is still a work in progress, but here is a snippet:

You can use vault to create reusable filters that will be optimised internally:
```js
const myFilterObject = vault.createFilter({
  type: 'Canvas',
  id: id => id.startsWidth('https://mydomain/'),
});
```

You can then add this filter as part of an index:
```js
const myIndex = vault.createIndex({
  name: 'My index',
  refresh: true,
  filter: myFilterObject,
  indexes: ['viewingHint'],
});
```

and this would be stored in the vault as something like this:
```js
const state = {
  indexes: [
    {
      filter: myFilterObject,
      index: {
        viewingHint: {
          individuals: ['http://mydomain/manifest-1/canvas-1.json'], 
          paged: ['http://mydomain/manifest-1/canvas-2.json', 'http://mydomain/manifest-1/canvas-3.json'],
        }
      }
    }
  ]
};
```

Then in your application, if you were to query the vault:
```js
const results = vault.query([myFilterObject, { 
  // This would be pre-filtered by the index.
  viewingHint: 'individuals',
  // Then this would be executed against the found objects.
  title: { $contains: 'Test' },
}]);
```

The initial set of results would be much smaller, and then more resource-intensive querying can be performed.


**It's very likely that users won't create these indexes themselves, although they could, but instead we wrap them inside of canvas panel when building our public APIs to optimise**

**Comment by tomcrane** [comment]
To clarify - at the beginning you have this example, from the original thinking:

```javascript
cp.query({
  type: "Annotation", // `DisplayAnnotation` here?
  motivation: "linking",
}).addEventListener("click", (e) => {
  // Handle the event here.
});
```

and then at the end you've reached

```javascript
const manifest = await vault.loadManifest( ... );
vault.addEventListener(manifest.items[0], 'onClick', () => {
  // Do something when clicking on the first canvas.
});
```

Would the first kind of usage still be supported? That is, hooking listeners to everything that matches a query?

**Reply by stephenwf** [reply]
Yes, the first is the nice abstraction, where you might be attaching to one, or two or 10 annotations. The vault API is concrete, you are attaching this event to this concrete thing.
